File: modules/rid-broadcast/ios/RidBroadcastModule.swift
// RidBroadcastModule.swift

import Foundation
import CoreBluetooth
import CryptoKit // Modern cryptography framework for signing

// The @objc attribute makes this Swift class available to the Objective-C runtime.
@objc(RidBroadcast)
class RidBroadcast: NSObject, CBPeripheralManagerDelegate {

  private var peripheralManager: CBPeripheralManager!
  private var broadcastTimer: Timer?
  private var allDroneData: [String: Any]?

  // ASTM F3411-22a Service UUID
  private let ridServiceUUID = CBUUID(string: "0000FFFA-0000-1000-8000-00805F9B34FB")

  override init() {
    super.init()
    // Initialize the peripheral manager on a background queue for performance.
    let queue = DispatchQueue(label: "com.yourprojectname.rid.bluetooth")
    peripheralManager = CBPeripheralManager(delegate: self, queue: queue)
  }

  // This tells React Native to initialize the module on the main thread.
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return true
  }

  /**
   * Starts a periodic broadcast, automatically cycling through different RID frames.
   * @param data A dictionary containing all necessary data for all frame types.
   * @param resolve The promise's resolve function.
   * @param reject The promise's reject function.
   */
  @objc(startBroadcast:resolver:rejecter:)
  func startBroadcast(data: [String: Any], resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    guard peripheralManager.state == .poweredOn else {
      reject("BLUETOOTH_OFF", "Bluetooth is not powered on.", nil)
      return
    }

    // Store the data and start the timer
    allDroneData = data
    stopBroadcastInternal() // Ensure any old timer is stopped

    // Use a Timer to call updateBroadcastFrame every second
    broadcastTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
        self?.updateBroadcastFrame()
    }
    
    // Run the timer on a background thread to avoid blocking the UI
    if let timer = broadcastTimer {
        RunLoop.current.add(timer, forMode: .common)
    }

    resolve(true)
  }
  
  /**
   * Stops the Bluetooth LE broadcast and cancels the timer.
   */
  @objc(stopBroadcast:rejecter:)
  func stopBroadcast(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    stopBroadcastInternal()
    resolve(true)
  }

  private func stopBroadcastInternal() {
    broadcastTimer?.invalidate()
    broadcastTimer = nil
    if peripheralManager.isAdvertising {
        peripheralManager.stopAdvertising()
    }
  }

  /// This function is called by the timer to generate and broadcast a new frame.
  private func updateBroadcastFrame() {
    guard let data = allDroneData, peripheralManager.state == .poweredOn else { return }

    let payload = createRemoteIdPayload(data: data)
    guard !payload.isEmpty else { return }

    let advertisementData: [String: Any] = [
        CBAdvertisementDataServiceUUIDsKey: [ridServiceUUID],
        CBAdvertisementDataServiceDataKey: [ridServiceUUID: payload]
    ]
    
    // To update advertising data, we must restart it.
    if peripheralManager.isAdvertising {
        peripheralManager.stopAdvertising()
    }
    peripheralManager.startAdvertising(advertisementData)
  }
  
  /// Creates a single RID payload frame based on the current time.
  private func createRemoteIdPayload(data: [String: Any]) -> Data {
    // Cycle through frames 0-5 using modulus division on the current time in seconds.
    let frameType = Int(Date().timeIntervalSince1970) % 6
    
    switch frameType {
      case 0: return createBasicIdPayload(data: data)
      case 1: return createLocationPayload(data: data)
      case 2: return createSelfIdPayload(data: data)
      case 3: return createSystemPayload(data: data)
      case 4: return createOperatorIdPayload(data: data)
      case 5: return createAuthenticationPayload(data: data)
      default: return Data()
    }
  }

  // MARK: - Payload Creation Functions

  private func createBasicIdPayload(data: [String: Any]) -> Data {
    guard let uasId = data["uasId"] as? String else { return Data() }
    var payload = Data()
    payload.append(0x00) // Message Type 0x0 (Basic ID), UAS ID Type 0 (Serial)
    payload.append(contentsOf: uasId.data(using: .ascii)!.prefix(20))
    return payload
  }

  private func createLocationPayload(data: [String: Any]) -> Data {
    var payload = Data()
    payload.append(0x10) // Message Type 0x1
    payload.append(UInt8((data["status"] as? Int) ?? 0))
    payload.append(int16: Int16((data["direction"] as? Double ?? 0.0) * 100))
    payload.append(int16: Int16((data["speedHorizontal"] as? Double ?? 0.0) * 100))
    payload.append(int16: Int16((data["speedVertical"] as? Double ?? 0.0) * 100))
    payload.append(int32: Int32((data["latitude"] as? Double ?? 0.0) * 1e7))
    payload.append(int32: Int32((data["longitude"] as? Double ?? 0.0) * 1e7))
    payload.append(int16: Int16((data["altitudePressure"] as? Double ?? 0.0) * 10))
    payload.append(int16: Int16((data["altitudeGeodetic"] as? Double ?? 0.0) * 10))
    payload.append(int16: Int16((data["height"] as? Double ?? 0.0) * 10))
    payload.append(contentsOf: [0x00, 0x00]) // Accuracies
    payload.append(int16: 0) // Timestamp
    return payload
  }
    
  private func createSelfIdPayload(data: [String: Any]) -> Data {
    guard let description = data["description"] as? String else { return Data() }
    var payload = Data()
    payload.append(0x20) // Message Type 0x2
    payload.append(0x01) // Description Type: Text
    payload.append(contentsOf: description.data(using: .ascii)!.prefix(23))
    return payload
  }

  private func createSystemPayload(data: [String: Any]) -> Data {
    var payload = Data()
    payload.append(0x30) // Message Type 0x3
    payload.append(UInt8((data["operatorLocationType"] as? Int) ?? 0))
    payload.append(int32: Int32((data["operatorLatitude"] as? Double ?? 0.0) * 1e7))
    payload.append(int32: Int32((data["operatorLongitude"] as? Double ?? 0.0) * 1e7))
    payload.append(int16: Int16((data["areaCount"] as? Int) ?? 0))
    payload.append(int32: Int32((data["areaRadius"] as? Int) ?? 0))
    payload.append(int32: Int32((data["areaCeiling"] as? Double ?? 0.0) * 10))
    payload.append(int32: Int32((data["areaFloor"] as? Double ?? 0.0) * 10))
    return payload
  }

  private func createOperatorIdPayload(data: [String: Any]) -> Data {
    guard let operatorId = data["operatorId"] as? String else { return Data() }
    var payload = Data()
    payload.append(0x40) // Message Type 0x4
    payload.append(0x00) // Operator ID Type: CAA
    payload.append(contentsOf: operatorId.data(using: .ascii)!.prefix(20))
    return payload
  }

  private func createAuthenticationPayload(data: [String: Any]) -> Data {
    guard let privateKeyString = data["privateKey"] as? String,
          let keyData = Data(base64Encoded: privateKeyString),
          let uasId = data["uasId"] as? String else {
      return createPlaceholderAuth()
    }
    
    do {
      // Use CryptoKit to create a private key and sign the data
      let privateKey = try P256.Signing.PrivateKey(derRepresentation: keyData)
      let messageToSign = "\(uasId):\(Int(Date().timeIntervalSince1970))"
      let messageData = messageToSign.data(using: .utf8)!
      let signature = try privateKey.signature(for: messageData)
      
      var payload = Data()
      payload.append(0x50) // Message Type 0x5
      payload.append(0x01) // Auth Type: Signature
      payload.append(signature.rawRepresentation.prefix(16)) // Truncate signature for this example
      return payload
    } catch {
      print("[RidBroadcast] Error creating auth payload: \(error)")
      return createPlaceholderAuth()
    }
  }

  private func createPlaceholderAuth() -> Data {
    return Data([0x50, 0x01] + [UInt8](repeating: 0xAA, count: 16))
  }

  // MARK: - CBPeripheralManagerDelegate

  func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
    if peripheral.state == .poweredOff {
      // If Bluetooth is turned off, stop everything.
      stopBroadcastInternal()
    }
    print("[RidBroadcast] Peripheral manager state changed to: \(peripheral.state.rawValue)")
  }
    
  func peripheralManagerDidStartAdvertising(_ peripheral: CBPeripheralManager, error: Error?) {
    if let error = error {
        print("[RidBroadcast] Failed to start advertising: \(error.localizedDescription)")
    }
  }
}

// Helper extension to append integers with correct byte order (little-endian)
extension Data {
    mutating func append(int16: Int16) {
        var value = int16.littleEndian
        append(Data(bytes: &value, count: MemoryLayout<Int16>.size))
    }
    mutating func append(int32: Int32) {
        var value = int32.littleEndian
        append(Data(bytes: &value, count: MemoryLayout<Int32>.size))
    }
}

File: modules/rid-broadcast-module/android/src/main/java/expo/modules/ridbroadcast/RidBroadcastModule.kt
package com.RidBroadcastModule

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothManager
import android.bluetooth.le.AdvertiseCallback
import android.bluetooth.le.AdvertiseData
import android.bluetooth.le.AdvertiseSettings
import android.bluetooth.le.BluetoothLeAdvertiser
import android.content.Context
import android.content.pm.PackageManager
import android.os.ParcelUuid
import android.util.Log
import com.facebook.react.bridge.*
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.StandardCharsets
import java.security.KeyFactory
import java.security.Signature
import java.security.spec.PKCS8EncodedKeySpec
import java.util.*

class RidBroadcastModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    private var bluetoothAdapter: BluetoothAdapter?
    private var bluetoothLeAdvertiser: BluetoothLeAdvertiser?
    private var isBroadcasting = false
    private var broadcastTimer: Timer? = null
    private var allDroneData: ReadableMap? = null

    // Remote ID Service UUID as per ASTM F3411-22a
    private val remoteIdServiceUuid = ParcelUuid.fromString("0000FFFA-0000-1000-8000-00805F9B34FB")

    init {
        val bluetoothManager = reactContext.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
        bluetoothAdapter = bluetoothManager.adapter
        bluetoothLeAdvertiser = bluetoothAdapter?.bluetoothLeAdvertiser
    }

    override fun getName(): String {
        return "RidBroadcastModule"
    }

    /**
     * Starts a periodic broadcast, automatically cycling through different RID frames.
     * @param data A map containing all necessary data for all frame types.
     * @param promise Resolves if successful, rejects on error.
     */
    @ReactMethod
    fun startBroadcast(data: ReadableMap, promise: Promise) {
        try {
            if (!isBluetoothSupported()) {
                promise.reject("BLUETOOTH_NOT_SUPPORTED", "Bluetooth LE is not supported on this device")
                return
            }

            if (bluetoothAdapter?.isEnabled != true) {
                promise.reject("BLUETOOTH_DISABLED", "Bluetooth is not enabled")
                return
            }

            allDroneData = data
            stopBroadcastInternal() // Ensure any old timer is stopped

            // Set up a new timer to call updateBroadcastFrame every second
            broadcastTimer = Timer()
            broadcastTimer?.scheduleAtFixedRate(object : TimerTask() {
                override fun run() {
                    updateBroadcastFrame()
                }
            }, 0, 1000)

            isBroadcasting = true
            promise.resolve(true)

        } catch (e: Exception) {
            promise.reject("START_BROADCAST_ERROR", e.message, e)
        }
    }

    /**
     * Stops the Bluetooth LE broadcast and cancels the timer.
     */
    @ReactMethod
    fun stopBroadcast(promise: Promise) {
        try {
            stopBroadcastInternal()
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject("STOP_BROADCAST_ERROR", e.message, e)
        }
    }

    private fun stopBroadcastInternal() {
        broadcastTimer?.cancel()
        broadcastTimer = null
        if (bluetoothLeAdvertiser != null && bluetoothAdapter?.isEnabled == true) {
             try {
                bluetoothLeAdvertiser?.stopAdvertising(advertiseCallback)
             } catch (e: SecurityException) {
                // Ignore if permissions are missing.
             }
        }
        isBroadcasting = false
    }

    /**
     * This function is called by the timer to generate and broadcast a new frame.
     */
    private fun updateBroadcastFrame() {
        if (bluetoothLeAdvertiser == null || allDroneData == null) return

        val payload = createRemoteIdPayload(allDroneData!!)
        if (payload.isEmpty()) return

        val settings = AdvertiseSettings.Builder()
            .setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_LOW_LATENCY)
            .setTxPowerLevel(AdvertiseSettings.ADVERTISE_TX_POWER_HIGH)
            .setConnectable(false)
            .setTimeout(0)
            .build()

        val advertiseData = AdvertiseData.Builder()
            .setIncludeDeviceName(false)
            .setIncludeTxPowerLevel(false)
            .addServiceUuid(remoteIdServiceUuid)
            .addServiceData(remoteIdServiceUuid, payload)
            .build()
        
        try {
            // To update data, we must stop the old advertisement and start a new one.
            bluetoothLeAdvertiser?.stopAdvertising(advertiseCallback)
            bluetoothLeAdvertiser?.startAdvertising(settings, advertiseData, advertiseCallback)
        } catch(e: SecurityException) {
            stopBroadcastInternal()
        }
    }


    /**
     * Creates a single RID payload frame based on the current time.
     * It cycles through the 6 main message types, one per second.
     */
    private fun createRemoteIdPayload(data: ReadableMap): ByteArray {
        // Use modulus division on the current time in seconds to cycle through frames 0-5.
        val frameType = (System.currentTimeMillis() / 1000) % 6

        return when (frameType) {
            0L -> createBasicIdPayload(data)
            1L -> createLocationPayload(data)
            2L -> createSelfIdPayload(data)
            3L -> createSystemPayload(data)
            4L -> createOperatorIdPayload(data)
            5L -> createAuthenticationPayload(data)
            else -> ByteArray(0)
        }
    }

    private fun createBasicIdPayload(data: ReadableMap): ByteArray {
        val buffer = ByteBuffer.allocate(25).order(ByteOrder.LITTLE_ENDIAN)
        val idType = 0 // UAS ID Type: Serial Number (0)
        val uasId = data.getString("uasId") ?: return ByteArray(0)
        buffer.put((0x0 shl 4 or idType).toByte()) // Message Type 0x0 (Basic ID)
        buffer.put(uasId.toByteArray(StandardCharsets.US_ASCII).take(20).toByteArray())
        return buffer.array().copyOf(buffer.position())
    }

    private fun createLocationPayload(data: ReadableMap): ByteArray {
        val buffer = ByteBuffer.allocate(25).order(ByteOrder.LITTLE_ENDIAN)
        buffer.put(0x10.toByte()) // Message Type 0x1 (Location/Vector)
        buffer.put((data.getInt("status")).toByte())
        buffer.putShort((data.getDouble("direction") * 100).toInt().toShort())
        buffer.putShort((data.getDouble("speedHorizontal") * 100).toInt().toShort())
        buffer.putShort((data.getDouble("speedVertical") * 100).toInt().toShort())
        buffer.putInt((data.getDouble("latitude") * 1e7).toInt())
        buffer.putInt((data.getDouble("longitude") * 1e7).toInt())
        buffer.putShort((data.getDouble("altitudePressure") * 10).toInt().toShort())
        buffer.putShort((data.getDouble("altitudeGeodetic") * 10).toInt().toShort())
        buffer.putShort((data.getDouble("height") * 10).toInt().toShort())
        buffer.put(0x00.toByte()) // Accuracies
        buffer.put(0x00.toByte()) // Accuracies
        buffer.putShort(0x0000.toShort()) // Timestamp
        return buffer.array().copyOf(buffer.position())
    }
    
    private fun createSelfIdPayload(data: ReadableMap): ByteArray {
        val buffer = ByteBuffer.allocate(25).order(ByteOrder.LITTLE_ENDIAN)
        val description = data.getString("description") ?: return ByteArray(0)
        buffer.put(0x20.toByte()) // Message Type 0x2 (Self-ID)
        buffer.put(0x01.toByte()) // Description Type: Text (1)
        buffer.put(description.toByteArray(StandardCharsets.US_ASCII).take(23).toByteArray())
        return buffer.array().copyOf(buffer.position())
    }

    private fun createSystemPayload(data: ReadableMap): ByteArray {
        val buffer = ByteBuffer.allocate(25).order(ByteOrder.LITTLE_ENDIAN)
        val operatorLocationType = data.getInt("operatorLocationType")
        buffer.put(0x30.toByte()) // Message Type 0x3 (System)
        buffer.put(operatorLocationType.toByte())
        buffer.putInt((data.getDouble("operatorLatitude") * 1e7).toInt())
        buffer.putInt((data.getDouble("operatorLongitude") * 1e7).toInt())
        buffer.putShort(data.getInt("areaCount").toShort())
        buffer.putInt(data.getInt("areaRadius"))
        buffer.putInt((data.getDouble("areaCeiling") * 10).toInt())
        buffer.putInt((data.getDouble("areaFloor") * 10).toInt())
        return buffer.array().copyOf(buffer.position())
    }

    private fun createOperatorIdPayload(data: ReadableMap): ByteArray {
        val buffer = ByteBuffer.allocate(25).order(ByteOrder.LITTLE_ENDIAN)
        val operatorId = data.getString("operatorId") ?: return ByteArray(0)
        buffer.put(0x40.toByte()) // Message Type 0x4 (Operator ID)
        buffer.put(0x00.toByte()) // Operator ID Type: CAA Issued (0)
        buffer.put(operatorId.toByteArray(StandardCharsets.US_ASCII).take(20).toByteArray())
        return buffer.array().copyOf(buffer.position())
    }

    private fun createAuthenticationPayload(data: ReadableMap): ByteArray {
        try {
            val privateKeyString = data.getString("privateKey")
            if (privateKeyString.isNullOrEmpty()) {
                return createPlaceholderAuth()
            }

            val keyBytes = Base64.getDecoder().decode(privateKeyString)
            val keySpec = PKCS8EncodedKeySpec(keyBytes)
            val keyFactory = KeyFactory.getInstance("EC")
            val privateKey = keyFactory.generatePrivate(keySpec)

            val uasId = data.getString("uasId") ?: ""
            val messageToSign = "$uasId:${System.currentTimeMillis() / 1000}"
            val messageBytes = messageToSign.toByteArray(StandardCharsets.UTF_8)

            val signature = Signature.getInstance("SHA256withECDSA")
            signature.initSign(privateKey)
            signature.update(messageBytes)
            val signatureBytes = signature.sign()

            val buffer = ByteBuffer.allocate(25).order(ByteOrder.LITTLE_ENDIAN)
            buffer.put(0x50.toByte()) // Message Type 0x5 (Authentication)
            buffer.put(0x01.toByte()) // Auth Type: Signature (1)
            buffer.put(signatureBytes.take(16).toByteArray())
            return buffer.array().copyOf(buffer.position())

        } catch (e: Exception) {
            Log.e("RidBroadcastModule", "Error creating authentication payload: ${e.message}")
            return createPlaceholderAuth()
        }
    }

    private fun createPlaceholderAuth(): ByteArray {
        val buffer = ByteBuffer.allocate(25).order(ByteOrder.LITTLE_ENDIAN)
        buffer.put(0x50.toByte())
        buffer.put(0x01.toByte())
        buffer.put(ByteArray(16))
        return buffer.array().copyOf(buffer.position())
    }

    private val advertiseCallback = object : AdvertiseCallback() {
        override fun onStartSuccess(settingsInEffect: AdvertiseSettings?) {
            super.onStartSuccess(settingsInEffect)
        }

        override fun onStartFailure(errorCode: Int) {
            super.onStartFailure(errorCode)
        }
    }
}
